### PR TITLE
AS-58 - Sites implementation

### PR DESCRIPTION
**Pull request by TomPed**
Added some more tests too. Still need to add to the test doc.

### CODE REVIEW

File: test/unit/sites.js [hunk 1]
```diff
@@ -3,179 +3,325 @@
...
+      });
+
+      promise.then(function(data) {
+        assert.deepEqual([site1], data);
         done();
       });
     });
 
-    xit('get a site', function(done) {
+    it('gets a site that does not exist', function(done) {
```
Comment (jacquelineali) [inline test/unit/sites.js]: Again, probably clearer if it says "Returns an empty array if the site_id doesn't exist"

File: test/unit/sites.js [hunk 2]
```diff
@@ -3,179 +3,325 @@
...
+        body: {}
+      })
+      .catch(function(err) {
+        assert.equal(err.status, 406);
+        assert.equal(err.message, 'Must provide site\'s name, or address');
+        done();
+      });
+    });
+
+    it('updates a site that does not exist', function(done) {
```
Comment (jacquelineali) [inline test/unit/sites.js]: More clear "It does nothing when asked to update a site that doesn't exist" or something

File: test/unit/sites.js [hunk 3]
```diff
@@ -3,179 +3,325 @@
...
+          }
+        });
+      })
+      .then(function(data) {
+        assert.deepEqual([], data);
+        done();
+      });
+    });
+
+    it('deletes a site that does not exist', function(done) {
```
Comment (jacquelineali) [inline test/unit/sites.js]: clarify: "Does nothing when told to delete a site that doesn't exist"
Or something like that

File: routes/sites.js [hunk 4]
```diff
@@ -1,25 +1,84 @@
...
 
+/**
+ * Get a sites with given id.
+ *
+ * @param {Object} req The given request object
+ * @return {Promise} The promise
+ */
 function getSite(req) {
-
+  return query('SELECT * FROM Site WHERE site_id = ' + req.params.site_id);
```
Comment (jacquelineali) [inline routes/sites.js]: If you have time though now, that'd be sick. Up to you though.

File: routes/sites.js [hunk 5]
```diff
@@ -1,25 +1,84 @@
...
+  return query('SELECT * FROM Site');
 }
 
+/**
+ * Creates a site.
+ *
+ * @param {Object} req The given request object
+ * @return {Promise} The promise
+ */
 function postSite(req) {
```
Comment (jacquelineali) [inline routes/sites.js]: This method should be named createSite(req) for consistency with the other endpoints. Don't forget to update that in the tests too! **Address before merging**

File: routes/sites.js [hunk 6]
```diff
@@ -1,25 +1,84 @@
...
+  if (!req.body || !req.body.site_name || !req.body.site_address) {
+    return Promise.reject({
+      status: 406,
+      message: 'Must provide site\'s name, and address'
+    });
+  }
 
+  return query('INSERT INTO Site (site_name, site_address) VALUES ("' + req.body.site_name + '", "' + req.body.site_address + '")');
```
Comment (jacquelineali) [inline routes/sites.js]: Note: What if we already have a site named "Northeastern International Village" with address "1155 Tremont Street?" If you don't check that it exists before posting, it'll create a duplicate with a different ID. Add a test case to catch this (e.g. don't create site if it already exists). You don't need to throw an error, but maybe just do nothing and log a message. **Address before merging**

File: test/unit/sites.js [hunk 7]
```diff
@@ -3,179 +3,325 @@
...
+
+      promise.then(function(data) {
+        assert.deepEqual([], data);
+        done();
+      });
+    });
+  });
+
+  describe('createSite(req)', function() {
+    it('create a site', function(done) {
```
Comment (jacquelineali) [inline test/unit/sites.js]: add the test mentioned above, for "don't create site if it already exists" **address before merging**

File: lib/utils.js [hunk 8]
```diff
@@ -64,6 +64,7 @@ function seed() {
       'INSERT INTO AcctToProgram (acct_id, program_id) VALUES (7, 1)',
       'INSERT INTO AcctToProgram (acct_id, program_id) VALUES (8, 2)',
       'INSERT INTO AcctToProgram (acct_id, program_id) VALUES (1, 1)',
+      'INSERT INTO AcctToProgram (acct_id, program_id) VALUES (1, 2)',
```
Comment (TomPed) [inline lib/utils.js]: I just needed to have some more data for my tests. Should not change anything with your tests.

File: routes/sites.js [hunk 9]
```diff
@@ -1,25 +1,84 @@
...
+const query = require('../lib/utils').query;
 
+/**
+ * Gets all sites.
+ *
+ * @param {Object} req The given request object
+ * @return {Promise} The promise
+ */
 function getSites(req) {
+  if (req.query && req.query.coach_id) {
```
Comment (TomPed) [inline routes/sites.js]: Nice catch. I just asked someone on the front-end team. They don't need any `GET`ters like that.

File: routes/sites.js (hunk 4 above)
Comment (TomPed) [inline routes/sites.js]: I dmed you. But I think that you `isValidDate` and `isPositiveInteger` should be in `utils.js` so everyone can use them. So I will not make this change now as we can open a PR later using your functions in `utils.js`! :smile:

File: test/unit/sites.js (hunk 7 above)
Comment (TomPed) [inline test/unit/sites.js]: Done.

**Comment by tkindy** [comment]
Deployment's currently broken. I'm working on it

**Pull request merged**